ADR 016: Removing Token Voting for Upgrades (#1491)
celestiaorg/celestia-app: docs/architecture/adr-016-social-upgrades.md

# ADR 016: Removing Token Voting for Upgrades

## Status

Accepted

## Changelog

- 2023/03/15: initial draft

## Context

> Standard cosmos-sdk based chains are sovereign and have the ability to hardfork, but the current implementation makes heavy use of onchain token voting. This bypasses social consensus, one of the core principles of Celestia. After a proposal is submitted and crosses some threshold of agreement, the halt height is set in the state machine for all nodes, and validators are expected to switch binaries after the state machine halts the node. The ultimate security, ruleset, and result of that ruleset rests solely on the shoulders of node operators, not validators or token holders. Therefore we are seeking a solution that engraves social consensus into the implementation itself.

Author: You are correct that people can always use which ever binary they choose, which is what the first part of the sentence was attempting to address. 
> Standard cosmos-sdk based chains are sovereign and have the ability to hardfork,

but the upgrade module does put a loose blocker to this by and the point of the onchain mechanism is to use token voting to decide. The social contract is to do whatever token voting determines, which is not the social contract that celestia wants to adhere to.

> Standard cosmos-sdk based chains are sovereign and have the ability to hardfork, but the current implementation makes heavy use of onchain token voting. This bypasses social consensus, one of the core principles of Celestia. After a proposal is submitted and crosses some threshold of agreement, the halt height is set in the state machine for all nodes, and validators are expected to switch binaries after the state machine halts the node. The ultimate security, ruleset, and result of that ruleset rests solely on the shoulders of node operators, not validators or token holders. Therefore we are seeking a solution that engraves social consensus into the implementation itself.

Author: updated, and I tried to focus on difference in social contract more

The most pertinent issue at the moment is that we are launching mainnet soon, and don't have a fully functional upgrade mechanism that empowers social consensus in place. We need to first decide how to remove token voting in a way that supports our future efforts to change the upgrade mechanism. The latter discussion on how to actually implement upgrades that fit all of our desired properties is out of scope for this ADR, and will be discussed separately.

> It should also be noted that this ADR is not discussion parameter changes, as limiting token voting on those is also a separate discussion.

Reviewer: ```suggestion
It should also be noted that this ADR is not discussing parameter changes, as limiting token voting on those is also a separate discussion.
```

## Alternative Approaches

### Option 1: Temporarily use the current token voting mechanism
Given that we don't have a lot of time until mainnet, we could leave the current implementation in place which gives us the option to use it if needed. Ideally, we would work on the longer term upgrade mechanism that respects social consensus and finish it before the first upgrade. While this option does allow for maximum flexibility, it is also very risky because if the mechanism is ever needed or used, then it could set a precedent for future upgrades.

### Option 2: Removal of token voting for upgrades

> The goal of this approach is to force social consensus to reach an upgrade instead of relying on token voting. It works by simply removing the current upgrade module. This way, the only way to upgrade the chain is to agree on the upgrade logic and the upgrade height.

Reviewer: Might be helpful to expand on what the upgrade mechanism would be. Given that we've decided to use a single binary sync which basically enables rolling upgrades, operators can update their binaries ahead of time and a height is embedded in the binary which indicates when to switch to the next state machine.

### Option 3: Adding a predetermined halt height (aka "difficulty bomb")
This option is not mutually exclusive to option 2. Its goal is to explicitly state that, without changing binaries, light clients will halt at a given height, despite what logic validators are running. This acts as an explicit statement to large token holders that they either come to some sort of agreement with the rest of the network, or chain will halt. Not coming to agreement is not a viable option.

## Decision

> TODO

Reviewer 2: I think it's safe to say we settled on Option 2

```toml
# HaltHeight contains a non-zero block height at which a node will gracefully
# halt and shutdown that can be used to assist upgrades and testing.
#
# Note: Commitment of state will be attempted on the corresponding block.
halt-height = 0

# HaltTime contains a non-zero minimum block time (in Unix seconds) at which
# a node will gracefully halt and shutdown that can be used to assist upgrades
# and testing.
#
# Note: Commitment of state will be attempted on the corresponding block.
halt-time = 0
```

## Consequences

If we adopt Option 2, then we will be able to remove token voting from the state machine sooner rather than later. This is riskier in that we will not have the battle tested mechanism, but it will force future upgrades that respect social consensus.

> ## References

Reviewer: Would be nice to reference the ADR018 as to a plan to upgrading

The most pertinent issue at the moment is that we are launching mainnet soon, and don't have a fully functional upgrade mechanism that empowers social consensus in place. We need to first decide how to remove token voting in a way that supports our future efforts to change the upgrade mechanism. The latter discussion on how to actually implement upgrades that fit all of our desired properties is out of scope for this ADR, and will be discussed separately.

It should also be noted that this ADR is not discussion parameter changes, as limiting token voting on those is also a separate discussion.

## Alternative Approaches

### Option 1: Temporarily use the current token voting mechanism

Given that we don't have a lot of time until mainnet, we could leave the current implementation in place which gives us the option to use it if needed. Ideally, we would work on the longer term upgrade mechanism that respects social consensus and finish it before the first upgrade. While this option does allow for maximum flexibility, it is also very risky because if the mechanism is ever needed or used, then it could set a precedent for future upgrades.

> ### Option 2: Predetermined halt height and removal of token voting for upgrades (aka "difficulty bomb")

Reviewer 3: [question] is there an Option 3 that removes token voting but doesn't introduce the difficulty bomb?

Reasons why a difficulty bomb may be unnecessary:
1. There isn't a need to introduce an upgrade mechanism prior to the next (currently unplanned) upgrade. Given the next upgrade isn't planned, we don't have a block height for it. Since no block height exists, we can't implement a difficulty bomb that coincides with when such an upgrade were to take place. Implementing a difficulty bomb with an estimated block height may result in several subsequent patches to delay the bomb.
2. The protocol may not need the difficulty bomb as an incentive to implement an upgrade mechanism. If the next upgrade brings improvements to the protocol, protocol participants may be sufficiently incentivized to introduce an upgrade mechanism to benefit from the proposed changes.

Author: we could yeah, but here I didn't want to specify all of the different combinations, as this was just a quick ADR to cover something that was discussed synchronously. 

> There isn't a need to introduce an upgrade mechanism prior to the next (currently unplanned) upgrade. Given the next upgrade isn't planned, we don't have a block height for it. Since no block height exists, we can't implement a difficulty bomb that coincides with when such an upgrade were to take place. Implementing a difficulty bomb with an estimated block height may result in several subsequent patches to delay the bomb.

Adding the difficulty bomb is inherently planning for an upgrade within some period of time. It's explicitly telling the validators that they have to at least make one change before that determined height.

> The protocol may not need the difficulty bomb as an incentive to implement an upgrade mechanism. If the next upgrade brings improvements to the protocol, protocol participants may be sufficiently incentivized to introduce an upgrade mechanism to benefit from the proposed changes.

The attack scenario is that the validators choose not to upgrade but social consensus does. This is just an explicit form of that. Participants could want to upgrade, but that does not mean validators will want to. I can add a line making this more explicit though!

Author: I'll add the third option though

```go
// SoftwareUpgrade implements the Msg/SoftwareUpgrade Msg service.
func (k msgServer) SoftwareUpgrade(goCtx context.Context, req *types.MsgSoftwareUpgrade) (*types.MsgSoftwareUpgradeResponse, error) {
   if k.authority != req.Authority {
       return nil, errors.Wrapf(gov.ErrInvalidSigner, "expected %s got %s", k.authority, req.Authority)
   }

   ctx := sdk.UnwrapSDKContext(goCtx)
   err := k.ScheduleUpgrade(ctx, req.Plan)
   if err != nil {
       return nil, err
   }

   return &types.MsgSoftwareUpgradeResponse{}, nil
}
```

#### Implement a deadline module into the state machine that will halt at a hardcoded height

There are a few different mechanisms that we could use to add a constant that
would halt all nodes in the network. This could be as simple as:

```go
func BeginBlocker(k keeper.Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
   if ctx.BlockHeader().Height >= BombHeight {
       panic("social consensus must be reached to pass the bomb height")
   }
   return
}
```

> We could also impose other similar restrictions elsewhere instead, such as voting to reject the block if the deadline height is `BombHeight` is reached.

Reviewer 3: ```suggestion
We could also impose other similar restrictions elsewhere instead, such as voting to reject the block if the deadline height in `BombHeight` is reached.
```

    }
    ...
    return nil
}
```

#### Halting the Node using Social Consensus

We hope to perform most upgrades using mechanism that doesn't involve shutting down and switching binaries, but depending on changes to the code, this might be difficult or not desirable (note that single binary syncing would still work fine). In that case, we would still require a mechanism to halt all nodes that are running the old binary in a way that respects social consensus. We can do that using the existing functionality in the application. Below are the configs in app.toml that would allow node operators to pick a height to shutdown their nodes at.

```toml
# HaltHeight contains a non-zero block height at which a node will gracefully
# halt and shutdown that can be used to assist upgrades and testing.
#
# Note: Commitment of state will be attempted on the corresponding block.
halt-height = 0

# HaltTime contains a non-zero minimum block time (in Unix seconds) at which
# a node will gracefully halt and shutdown that can be used to assist upgrades
# and testing.
#
# Note: Commitment of state will be attempted on the corresponding block.
halt-time = 0
```

## Consequences

> If we adopt Option 2, then we will be able to remove token voting from the statemachine sooner rather than later. This is riskier in that we will not have the battle tested mechanism, but it will force future upgrades that respect social consensus.

Reviewer 3: ```suggestion
If we adopt Option 2, then we will be able to remove token voting from the state machine sooner rather than later. This is riskier in that we will not have the battle tested mechanism, but it will force future upgrades that respect social consensus.
```

## Alternative Approaches

### Option 1: Temporarily use the current token voting mechanism
Given that we don't have a lot of time until mainnet, we could leave the current implementation in place which gives us the option to use it if needed. Ideally, we would work on the longer term upgrade mechanism that respects social consensus and finish it before the first upgrade. While this option does allow for maximum flexibility, it is also very risky because if the mechanism is ever needed or used, then it could set a precedent for future upgrades.

### Option 2: Removal of token voting for upgrades

> The goal of this approach is to force social consensus to reach an upgrade instead of relying on token voting. It works by simply removing the currently upgrade module. This way, the only way to upgrade the chain is to agree on the upgrade logic and the upgrade height.

Reviewer 3: ```suggestion
The goal of this approach is to force social consensus to reach an upgrade instead of relying on token voting. It works by simply removing the current upgrade module. This way, the only way to upgrade the chain is to agree on the upgrade logic and the upgrade height.
```
